### PR TITLE
Update to NDK 27

### LINKED ISSUE #20927: LTS NDK 27 was just released
[We should update to use it](https://developer.android.com/ndk/downloads). We can finally get rid of `-fno-openmp-implicit-rpath`, and I'm sure other issues too.

### PR DESCRIPTION
**Pull request by licy183**
Closes #20927

I'll update the flang toolchain this weekend.

### CODE REVIEW

**Review by Grimler91 — APPROVED** [review]

**Comment by licy183** [comment]
Emmm... Seems out of space on GitHub Action...

**Comment by Biswa96** [comment]
Some ideas.
1. In packages workflow file, move the free addition disk space section when container image is built.
2. Or, move the package update in separate commit.
3. Or, merge this. There is nothing to fail.

**Comment by licy183** [comment]
Should wait for https://github.com/actions/runner-images/issues/10314, otherwise CI will fail on the packages on `big-pkg.list`.

**Comment by Biswa96** [comment]
> ...otherwise CI will fail on the packages in `big-pkg.list`.

It could be installed with sdkmanager command.

**Comment by truboxl** [comment]
Can `-fopenmp -static-openmp` be dropped?

**Comment by finagolfin** [comment]
> Can -fopenmp -static-openmp be dropped?

I think some packages require it, so not dropped completely, but if you want to remove it from this central location, see which packages fail to build or run, then add it for just those, that's fine.

**Comment by truboxl** [comment]
Ok I will kick off a test build with NDK r26b with the openmp options removed...

**Comment by IntinteDAO** [comment]
> Closes #20927
> 
> I'll update the flang toolchain this weekend.

Why can't we just update to ndk 29?

**Comment by TomJo2000** [comment]
> Why can't we just update to ndk 29?

I think you are confusing NDK version and API level.

**Comment by IntinteDAO** [comment]
> > Why can't we just update to ndk 29?
> 
> I think you are confusing NDK version and API level.

Possible. I just tried to build one app on Termux and it throws an error that the function appeared in Android 29. I assumed it was about NDK because it's a C++ app

**Comment by TomJo2000** [comment]
Yeah you're probably talking about the API target.
See: 
- termux/termux-exec#24 and
- termux/termux-app#2155

for additional information.

File: scripts/setup-android-sdk.sh
Comment (finagolfin) [inline scripts/setup-android-sdk.sh:16]: Any reason this pull keeps 23c around, rather than switching to keeping 26b as the backup and 27 as the new default?

**Comment by finagolfin** [comment]
> Possible. I just tried to build one app on Termux and it throws an error that the function appeared in Android 29.

Android 29 is the version of the OS APIs, while NDK 27 is a new development toolchain that supports [Android OS APIs 21 through 35](https://en.m.wikipedia.org/wiki/Android_version_history), the latest OS API version that will be released next month.

We target Android OS API 24, partially for the reasons given in the links @TomJo2000 gave.

File: scripts/setup-android-sdk.sh
Comment (truboxl) [reply scripts/setup-android-sdk.sh:16]: It is the last NDK to support `-march=armv7-a -mfpu=vfpv3-d16`

Comment (finagolfin) [reply scripts/setup-android-sdk.sh:16]: Some packages are still using it for that?

Comment (truboxl) [reply scripts/setup-android-sdk.sh:16]: All existing packages are built with LDFLAGS `-mfpu=neon` so no. This is only kept to extend hardware support for Termux. I am pretty much the only maintainer that still uses it.

Comment (finagolfin) [reply scripts/setup-android-sdk.sh:16]: OK, what do you use 23c for, some old armv7 devices?

@licy183, would it make sense to keep 26b around also as a backup or no need?

Comment (licy183) [reply scripts/setup-android-sdk.sh:16]: Emmm... I think there is no need to keep NDK 26b.

File: ndk-patches/27/stdio.h.patch
Comment (finagolfin) [inline ndk-patches/27/stdio.h.patch:7]: Needed this last commit for the upcoming Swift 6, shouldn't affect anything else.

Comment (Biswa96) [reply ndk-patches/27/stdio.h.patch:7]: Could you explain the reason for this change or what issues are fixed?

**Comment by truboxl** [comment]
> Ok I will kick off a test build with NDK r26b with the openmp options removed...

I have yet to see any build failure when openmp options are removed (currently up to 832 packages)

**Comment by finagolfin** [comment]
@Grimler91, do we need a full package rebuild after this?

File: ndk-patches/27/stdio.h.patch
Comment (finagolfin) [reply ndk-patches/27/stdio.h.patch:7]: Sure, wasn't sure if anyone would be interested. 😄

Swift 6 adds [a fully modularized `android.modulemap`](https://github.com/swiftlang/swift/pull/72161/files#diff-1fd3199a3e1289fac4557405173ac51a19939b34f76f89676fa38833761398bd), as opposed to [the `glibc.modulemap`](https://github.com/swiftlang/swift/blob/main/stdlib/public/Platform/glibc.modulemap.gyb) we reuse from linux in the current Swift 5.10 package. This change is primarily to help with [the new C++ interoperability features of Swift](https://www.swift.org/documentation/cxx-interop/), which require a modularized libc module map to go with [the libc++ module.modulemap from the NDK that we currently ship](https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/refs/heads/main/clang-r522817/include/c++/v1/module.modulemap).

The problem is these additional headers we patch in for Termux break that modularity, in ways I don't understand. So, I found that [the Swift compiler sets `__swift__` when importing these headers](https://github.com/swiftlang/swift/blob/main/lib/ClangImporter/ClangImporter.cpp#L510) and undefined these when building for Swift, including the `ctermid()` inline function definition that needs that header.

File: scripts/setup-android-sdk.sh
Comment (Grimler91) [reply scripts/setup-android-sdk.sh:16]: Wasn't 23c useful for compiling some packages in termux-user-repository as well?

**Comment by Grimler91** [comment]
@finagolfin based on previous updates there should be no need. Also doesn't seem to be any large obvious changes that should affect us in [changelog](https://github.com/android/ndk/wiki/Changelog-r27), but I'll try to run some builds to check for new errors affecting packages

File: ndk-patches/27/stdlib.h.patch
Comment (Grimler91) [inline ndk-patches/27/stdlib.h.patch:9]: Missing `#` here, `#endif`

Comment (finagolfin) [reply ndk-patches/27/stdlib.h.patch:9]: Thanks

Comment (finagolfin) [reply ndk-patches/27/stdlib.h.patch]: @licy183, I didn't pull again before fixing this typo just now. Feel free to push your local branch again with my last commit added if I missed something.

**Comment by finagolfin** [comment]
I'm going to test my daily Swift CI for Android now with NDK 27, will add that cached NDK download step to this pull too if all goes well, so we won't have to wait on the github runner image to add NDK 27.

**Comment by Grimler91** [comment]
Busybox fails to build due to missing symbols in static lib generation, so maybe several packages that internally use static libs will have similar errors:

```
networking/tc.c:236:20  AR      procps/lib.a
: error: use of undeclared identifier 'TCA_CBQ_MAX'
  236 |         struct rtattr *tb[TCA_CBQ_MAX+1];
      |                           ^
networking/tc.c:247:26: error: use of undeclared identifier 'TCA_CBQ_MAX'
  247 |         parse_rtattr_nested(tb, TCA_CBQ_MAX, opt);
      |                                 ^
  CC      coreutils/mktemp.o
networking/tc.c:249:9: error: use of undeclared identifier 'TCA_CBQ_RATE'
  249 |         if (tb[TCA_CBQ_RATE]) {
      |                ^
networking/tc.c:250:22: error: use of undeclared identifier 'TCA_CBQ_RATE'
  250 |                 if (RTA_PAYLOAD(tb[TCA_CBQ_RATE]) < sizeof(*r))
      |                                    ^
networking/tc.c:253:20: error: use of undeclared identifier 'TCA_CBQ_RATE'
  253 |                         r = RTA_DATA(tb[TCA_CBQ_RATE]);
      |                                         ^
networking/tc.c:255:9: error: use of undeclared identifier 'TCA_CBQ_LSSOPT'
  255 |         if (tb[TCA_CBQ_LSSOPT]) {
      |                ^
networking/tc.c:256:22: error: use of undeclared identifier 'TCA_CBQ_LSSOPT'
  256 |                 if (RTA_PAYLOAD(tb[TCA_CBQ_LSSOPT]) < sizeof(*lss))
      |                                    ^
networking/tc.c:256:47: error: invalid application of 'sizeof' to an incomplete type 'struct tc_cbq_lssopt'
  256 |                 if (RTA_PAYLOAD(tb[TCA_CBQ_LSSOPT]) < sizeof(*lss))
      |                                                             ^~~~~~
networking/tc.c:238:9: note: forward declaration of 'struct tc_cbq_lssopt'
  238 |         struct tc_cbq_lssopt *lss = NULL;
      |                ^
networking/tc.c:259:22: error: use of undeclared identifier 'TCA_CBQ_LSSOPT'
  259 |                         lss = RTA_DATA(tb[TCA_CBQ_LSSOPT]);
      |                                           ^
networking/tc.c:261:9: error: use of undeclared identifier 'TCA_CBQ_WRROPT'
  261 |         if (tb[TCA_CBQ_WRROPT]) {
      |                ^
```

**Comment by Biswa96** [comment]
That's probably not related to NDK update, see https://www.mail-archive.com/busybox%40busybox.net/msg29286.html. There are similar issues reported about glibc or Linux kernel update (googlefu).

**Comment by finagolfin** [comment]
I'm seeing an issue with using `CMAKE_SYSTEM_PROCESSOR` with NDK 27, which doesn't affect Termux because the standalone toolchain does not pull in the NDK's CMake config. I'll ask them what's going on, but shouldn't be a stopper for us.

**Comment by finagolfin** [comment]
Heh, I first looked up if anyone had reported a `CMAKE_SYSTEM_PROCESSOR` issue with the NDK, and of course, [the only one was a Termux user years ago about a different problem](https://github.com/android/ndk/issues?q=is%3Aissue+cmake_system_processor). 😄

**Comment by Grimler91** [comment]
> That's probably not related to NDK update

Probably due to new llvm version in ndk r27. It builds fine with ndk r26b.

File: scripts/setup-android-sdk.sh
Comment (finagolfin) [reply scripts/setup-android-sdk.sh:16]: Hmm, github is removing everything older than NDK 26 from their CI images, actions/runner-images#10342.

**Comment by Biswa96** [comment]
> Probably due to new llvm version in ndk r27. It builds fine with ndk r26b.

TCA_CBQ_WRROPT was removed in this bionic commit https://android.googlesource.com/platform/bionic/+/b830ddf90e70702851d27f76b9beff472f5ff9a1%5E%21/#F51 due to the Linux headers change as mentioned in that busybox mail.

**Comment by truboxl** [comment]
I completed test build all the packages in main, root, x11 repo. No openmp related build issue found when remove the openmp options and using NDK r26b. There are some undiscovered build issues that did pop up. Will file a tracker issue or individual issue for that.

**Comment by truboxl** [comment]
There is still flags injected in the form pc files. Likely will show up as `clang: error: unknown argument: '-fno-openmp-implicit-rpath'` even if removed in setup_toolchain.
```console
$ grep fopenmp -nHR /data/data/com.termux/files/usr/lib/pkgconfig/
/data/data/com.termux/files/usr/lib/pkgconfig/ncursesw.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
/data/data/com.termux/files/usr/lib/pkgconfig/curses.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
/data/data/com.termux/files/usr/lib/pkgconfig/ncurses.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
/data/data/com.termux/files/usr/lib/pkgconfig/termcap.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
/data/data/com.termux/files/usr/lib/pkgconfig/tinfo.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
/data/data/com.termux/files/usr/lib/pkgconfig/tic.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
/data/data/com.termux/files/usr/lib/pkgconfig/menuw.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lmenuw
/data/data/com.termux/files/usr/lib/pkgconfig/panelw.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lpanelw
/data/data/com.termux/files/usr/lib/pkgconfig/menu.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lmenuw
/data/data/com.termux/files/usr/lib/pkgconfig/panel.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lpanelw
/data/data/com.termux/files/usr/lib/pkgconfig/ruby-3.3.pc:51:DLDFLAGS=-L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -Wl,--as-needed -Wl,-z,relro,-z,now -Wl,--no-as-needed,-landroid-support,--as-needed -Wl,--compress-debug-sections=zlib
/data/data/com.termux/files/usr/lib/pkgconfig/libldns.pc:12:Libs.private: -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -Wl,--enable-new-dtags -Wl,--as-needed -Wl,-z,relro,-z,now
/data/data/com.termux/files/usr/lib/pkgconfig/ldns.pc:12:Libs.private: -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -Wl,--enable-new-dtags -Wl,--as-needed -Wl,-z,relro,-z,now
/data/data/com.termux/files/usr/lib/pkgconfig/form.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lformw
/data/data/com.termux/files/usr/lib/pkgconfig/formw.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lformw
```

**Comment by finagolfin** [comment]
> There is still flags injected in the form pc files. Likely will show up as clang: error: unknown argument: '-fno-openmp-implicit-rpath' even if removed in setup_toolchain.

Those are coming from upstream or were inserted by our Termux builds? Either way, it should be easy to remove that flag.

File: .github/workflows/packages.yml
Comment (truboxl) [inline .github/workflows/packages.yml:203]: termux_setup_ninja not good?

Comment (finagolfin) [reply .github/workflows/packages.yml:205]: This is much simpler, though I suppose more likely to get out of date?

**Comment by finagolfin** [comment]
Alright, added a commit to download NDK 27 to the github CI for big packages and cache it for later, plus kicked off a build of libllvm using it.

**Comment by finagolfin** [comment]
Libllvm setup with the new NDK all looked fine, but failed only because I didn't update the flang version also, which is enforced to match. Kicked off a new build of zig instead, which won't take hours like libllvm.

**Comment by finagolfin** [comment]
Alright, Zig built fine with NDK 27. I didn't know it shipped with its own copy of the LLVM source now, so that builds fine too.

This pull now works for all packages and can be used as a base for any further CI testing wanted, before merging.

**Comment by truboxl** [comment]
NDK r27 fails to build current apt

**Comment by Grimler91** [comment]
> NDK r27 fails to build current apt

Can reproduce as well, error looks something like:

```
In file included from /home/builder/.termux-build/apt/src/apt-pkg/depcache.cc:12:
In file included from /home/builder/.termux-build/apt/build/include/apt-pkg/algorithms.h:32:
In file included from /home/builder/.termux-build/apt/build/include/apt-pkg/cachefilter.h:9:
In file included from /home/builder/.termux-build/apt/build/include/apt-pkg/pkgcache.h:78:
In file included from /home/builder/.termux-build/apt/build/include/apt-pkg/mmap.h:31:
In file included from /home/builder/.termux-build/_cache/android-r27-api-24-v0/sysroot/usr/include/c++/v1/string:625:
In file included from /home/builder/.termux-build/_cache/android-r27-api-24-v0/sysroot/usr/include/c++/v1/string_view:936:
In file included from /home/builder/.termux-build/_cache/android-r27-api-24-v0/sysroot/usr/include/c++/v1/algorithm:1847:
In file included from /home/builder/.termux-build/_cache/android-r27-api-24-v0/sysroot/usr/include/c++/v1/__algorithm/nth_element.h:15:
/home/builder/.termux-build/_cache/android-r27-api-24-v0/sysroot/usr/include/c++/v1/__algorithm/sort.h:319:44: error: invalid operands to binary expression ('const APT::Container_iterator<APT::VersionContainerInterface, std::vector<pkgCache::VerIterator>, APT::VersionContainer<std::vector<pkgCache::VerIterator>>>' and 'difference_type' (aka 'long'))
  319 |   for (_RandomAccessIterator __i = __first + difference_type(1); __i != __last; ++__i) {
      |                                    ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~
/home/builder/.termux-build/_cache/android-r27-api-24-v0/sysroot/usr/include/c++/v1/__iterator/reverse_iterator.h:302:1: note: candidate template ignored: could not match 'reverse_iterator<_Iter>' against 'difference_type' (aka 'long')
  302 | operator+(typename reverse_iterator<_Iter>::difference_type __n, const reverse_iterator<_Iter>& __x) {
      | ^
/home/builder/.termux-build/_cache/android-r27-api-24-v0/sysroot/usr/include/c++/v1/__iterator/wrap_iter.h:194:1: note: candidate template ignored: could not match '__wrap_iter<_Iter1>' against 'difference_type' (aka 'long')
  194 | operator+(typename __wrap_iter<_Iter1>::difference_type __n, __wrap_iter<_Iter1> __x) _NOEXCEPT {
      | ^
[ ... ]
```

**Comment by truboxl** [comment]
> Those are coming from upstream or were inserted by our Termux builds? Either way, it should be easy to remove that flag.

Termux builds. Unfortunately "too many" ncurses based packages fail to build on NDK r27. I think "all" the OpenMP options should be dropped and re-enabled per package.

The current sox package that was enabled together with toolchain wide OpenMP options in #6146 just dont expose `--multi-threaded` option anymore nor does `$PREFIX/lib/libsox.so.3` has any OpenMP symbols. Both of these are true on Ubuntu. I am not going to check back whether is truly confirmed working back then or because of NDK upgrade stopped working since then.

Here is what I am thinking to do:
1. Drop all the OpenMP build options
2. Investigate sox and enable OpenMP properly
3. Write tests to check for OpenMP symbols when `-fopenmp` is specified, also fail build if `libomp.so` is added as dependency
4. Rebuild ncurses

However whether to do this before or after switching to NDK r27 is up for debate

**Comment by finagolfin** [comment]
Updated this pull to use NDK 27 from the CI image for non-Docker builds, now that it's deployed.

> whether to do this before or after switching to NDK r27 is up for debate

After the switch IMO, as we should get this in soon.

File: scripts/setup-android-sdk.sh
Comment (codenamedpkt) [reply scripts/setup-android-sdk.sh:16]: Time to drop out support of `armv7-a`, `aarch64` aka `armv8` seem common now. We need support `riscv64` and `riscv128` in long future.

**Comment by finagolfin** [comment]
Time to get this in? I've moved my Swift CI for Android to NDK 27 without a problem.

**Review by truboxl — APPROVED** [review]

**Comment by licy183** [comment]
Tests fine on all of my devices (Android9-MIX2s-arm, Android13-MIX4-aarch64, Android7-Bluestacks-i686, Android7-Bluestacks-x86_64). I'll merge this now.

**Comment by pgaskin** [comment]
>     /data/data/com.termux/files/usr/lib/pkgconfig/ncursesw.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
>     /data/data/com.termux/files/usr/lib/pkgconfig/curses.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
>     /data/data/com.termux/files/usr/lib/pkgconfig/ncurses.pc:17:Libs:  -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -fno-openmp-implicit-rpath -Wl,--enable-new-dtags -lncursesw
>
> Those are coming from upstream or were inserted by our Termux builds?

They were added by the Termux build. They weren't present yesterday.

**Comment by licy183** [comment]
> They were added by the Termux build. They weren't present yesterday.

It will be fixed in #21038.

**Pull request merged**
